scripts: build: syscall and subsystem metadata are emitted to JSON even when commented out

### Describe the bug

When I was working on an out‑of‑tree device driver I had to comment out several sections of code. I then discovered that, even for commented‑out code, the *syscall* and *subsystem* metadata are still emitted to the generated JSON file. This becomes especially problematic when a reference syscall implementation is copied and its name is not changed: the build finishes with a large number of warnings, e.g.:

```
zephyr/include/generated/zephyr/syscall_list.h:202:9: warning: "K_SYSCALL_FUNC" redefined
  202 | #define K_SYSCALL_FUNC 196
      |         ^~~~~~~~~~~~~~~~
```

The warnings appear even though the offending syscall is surrounded by comments.

**Proposed fix**

To avoid this situation we should *strip comments* from the source files **before** the parser extracts syscall and subsystem metadata. The ideal way would be to let the C pre‑processor handle this, but the build system does not invoke the pre‑processor for this step. Therefore the comment removal can be performed directly inside `scripts/build/parse_syscalls.py`, after which the filtered source can be processed with the existing regular‑expression based parser.

### Regression

- [ ] This is a regression.

### Steps to reproduce

_No response_

### Relevant log output

```shell

```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

_No response_

### Additional Context

_No response_

--- a/scripts/build/parse_syscalls.py
+++ b/scripts/build/parse_syscalls.py
@@ -53,11 +53,37 @@ struct\s+                       # struct keyword is next
 [{]                             # Open curly bracket
 '''
 
+# Matches C/C++ string or character literals (single‑ or double‑quoted),
+# correctly handling escaped quotes and backslashes.
+c_string_pattern = r'(?P<string>"(?:\\.|[^"\\])*"|\'(?:\\.|[^\'\\])*\')'
+# Matches C/C++ comments: either a block comment /* … */ (spanning any number
+# of lines) or a line comment // … that ends at the newline (with optional \r)
+c_comment_pattern = r"(?P<comment>/\*.*?\*/|//[^\n]*\r?$)"
+c_string_or_comment_regex = re.compile(
+    c_string_pattern + r'|' + c_comment_pattern, re.MULTILINE | re.DOTALL
+)
+
 
 def tagged_struct_update(target_list, tag, contents):
     regex = re.compile(tagged_struct_decl_template % tag, regex_flags)
     items = [mo.groups()[0].strip() for mo in regex.finditer(contents)]
     target_list.extend(items)
+
+
+def remove_c_comments(string):
+    """
+    Remove C style comments from input string
+    """
+
+    def _replace_match(m):
+        if m.lastgroup != "comment":
+            return m.group()
+        # Replace all non-newline characters in comments with spaces,
+        # preserving newlines to avoid merging lines.
+        text = m.group()
+        return "".join(ch if ch in ("\n", "\r") else " " for ch in text)
+
+    return c_string_or_comment_regex.sub(_replace_match, string)
 
 
 def analyze_headers(include_dir, scan_dir, file_list):
@@ -129,10 +155,13 @@ def analyze_headers(include_dir, scan_dir, file_list):
 
         try:
             to_emit = syscall_files[one_file]["emit"] | args.emit_all_syscalls
+            contents_no_comments = remove_c_comments(contents)
 
-            syscall_result = [(mo.groups(), fn, to_emit) for mo in syscall_regex.finditer(contents)]
+            syscall_result = [
+                (mo.groups(), fn, to_emit) for mo in syscall_regex.finditer(contents_no_comments)
+            ]
             for tag in struct_tags:
-                tagged_struct_update(tagged_ret[tag], tag, contents)
+                tagged_struct_update(tagged_ret[tag], tag, contents_no_comments)
         except Exception as e:
             sys.stderr.write(f"While parsing {fn}\n")
             raise e

